src/Tablebot/Plugins/Roll/Dice/DiceEval.hs

```diff
@@ -26,47 +27,75 @@ import Tablebot.Utility.Discord (Format (..), formatInput, formatText)
 import Tablebot.Utility.Exception (BotException (EvaluationException), catchBot, throwBot)
 import Tablebot.Utility.Random (chooseOne)
 
--- | A wrapper type to differentiate between the RNGCount and other Integers.
+-- | A wrapper type to differentiate between the RNGCount and other Integers,
+-- as well as store variables throughout the program.
 --
 -- Represents the total number of calls to the RNG throughout the program
 -- (effectively, how many die rolls have occured).
-newtype RNGCount = RNGCount {getRNGCount :: Integer} deriving (Eq, Ord)
+data ProgramState = ProgramState
+  { getRNGCount :: Integer,
+    getVariables :: Map Text (Either ListValues Expr)
+  }
+  deriving (Show)
+
+startState :: ProgramState
+startState = ProgramState 0 empty
+
+type ProgramStateM = StateT ProgramState IO
+
+-- | Add the given variable to the `ProgramState`
+addVariable :: Text -> Either ListValues Expr -> ProgramStateM ()
+addVariable t val = modify $ \s -> s {getVariables = M.insert t val (getVariables s)}
 
 -- | The maximum depth that should be permitted. Used to limit number of dice
 -- and rerolls.
-maximumRNG :: RNGCount
-maximumRNG = RNGCount 150
+maximumRNG :: Integer
+maximumRNG = 150
 
 maximumListLength :: Integer
 maximumListLength = 50
 
 -- | Increment the rngcount by 1.
-incRNGCount :: RNGCount -> RNGCount
-incRNGCount (RNGCount i) = RNGCount (i + 1)
+incRNGCount :: ProgramStateM ()
+incRNGCount = modify (\s -> s {getRNGCount = getRNGCount s + 1}) >> checkRNGCount
 
 -- | Check whether the RNG count has been exceeded by the integer given.
-checkRNGCount :: RNGCount -> IO ()
-checkRNGCount i =
-  when (i > maximumRNG) $ throwBot $ EvaluationException ("exceeded maximum rng count (" <> show (getRNGCount maximumRNG) <> ")") []
+checkRNGCount :: ProgramStateM ()
+checkRNGCount = do
+  rngCount <- gets getRNGCount
+  when (rngCount > maximumRNG) $ evaluationException ("Maximum RNG count exceeded (" <> pack (show maximumRNG) <> ")") []
 
 -- | Utility function to throw an `EvaluationException` when using `Text`.
 evaluationException :: (MonadException m) => Text -> [Text] -> m a
 evaluationException nm locs = throwBot $ EvaluationException (unpack nm) (unpack <$> locs)
 
 --- Evaluating an expression. Uses IO because dice are random
 
+-- | Evaluating a full program
+evalProgram :: Program -> IO (Either [(Integer, Text)] Integer, Text)
+evalProgram (Program ss elve) =
+  evalStateT
+    ( do
+        t <- foldl' (\b s -> b >>= \t -> evalStatement s >>= \st -> return (t <> st)) (return "") ss
```

Comment:
I'd want either a better variable name here or a comment explaining, particularly with the weird overshadowing of `t` going on within this line. I _think_ this is going through each statement in turn but I'd rather the code explained that.